File: app/auth/firebase_auth.py
import os
import json
import logging
from datetime import datetime, timedelta
from typing import Optional, Dict, Any

import firebase_admin
from firebase_admin import auth, credentials
import jwt

logger = logging.getLogger(__name__)
logging.basicConfig(level=logging.INFO)


class FirebaseAuthService:
    def __init__(self):
        self._initialize_firebase()
        self.jwt_secret = os.getenv("JWT_SECRET", "your-secret-key")
        self.jwt_algorithm = "HS256"
        self.access_token_expiry = timedelta(hours=1)
        self.refresh_token_expiry = timedelta(days=7)

    def _initialize_firebase(self):
        """Initialize Firebase Admin SDK using env credentials, service file, or default."""
        try:
            firebase_credentials = os.getenv("FIREBASE_CREDENTIALS")
            if firebase_credentials:
                cred = credentials.Certificate(json.loads(firebase_credentials))
            else:
                service_account_path = os.getenv("FIREBASE_SERVICE_ACCOUNT_PATH")
                if service_account_path and os.path.exists(service_account_path):
                    cred = credentials.Certificate(service_account_path)
                else:
                    cred = credentials.ApplicationDefault()

            firebase_admin.initialize_app(cred)
            logger.info("Firebase initialized successfully.")
        except Exception as e:
            logger.error(f"Firebase initialization error: {e}")
            raise RuntimeError("Failed to initialize Firebase SDK.") from e

    async def create_user(self, email: str, password: str, first_name: str, last_name: str) -> Dict[str, Any]:
        """Create a new Firebase user with custom claims."""
        try:
            user_record = auth.create_user(
                email=email,
                password=password,
                display_name=f"{first_name} {last_name}",
                email_verified=False
            )
            auth.set_custom_user_claims(user_record.uid, {
                "first_name": first_name,
                "last_name": last_name,
                "role": "user"
            })

            return {
                "id": user_record.uid,
                "email": user_record.email,
                "first_name": first_name,
                "last_name": last_name,
                "is_active": not user_record.disabled,
                "created_at": str(user_record.user_metadata.creation_timestamp)
            }
        except Exception as e:
            logger.error(f"User creation failed: {e}")
            raise RuntimeError(f"User creation failed: {e}") from e

    async def sign_in_user(self, email: str, password: str) -> Dict[str, Any]:
        """
        Simulate sign-in (Firebase doesn't allow password verification via Admin SDK).
        Production use should use Firebase Auth REST API.
        """
        try:
            user_record = auth.get_user_by_email(email)

            if user_record.disabled:
                raise PermissionError("User account is disabled")

            access_token = self._generate_access_token(user_record.uid, user_record.email)
            refresh_token = self._generate_refresh_token(user_record.uid)
            custom_claims = auth.get_custom_user_claims(user_record.uid) or {}

            return {
                "access_token": access_token,
                "refresh_token": refresh_token,
                "user": {
                    "id": user_record.uid,
                    "email": user_record.email,
                    "first_name": custom_claims.get("first_name", ""),
                    "last_name": custom_claims.get("last_name", ""),
                    "is_active": not user_record.disabled,
                    "created_at": str(user_record.user_metadata.creation_timestamp)
                }
            }
        except Exception as e:
            logger.error(f"Sign-in failed: {e}")
            raise RuntimeError("Authentication failed.") from e

    async def verify_token(self, token: str) -> Optional[Dict[str, Any]]:
        """Verify Firebase ID token and return user metadata."""
        try:
            decoded = auth.verify_id_token(token)
            user_record = auth.get_user(decoded["uid"])
            custom_claims = auth.get_custom_user_claims(user_record.uid) or {}

            return {
                "uid": user_record.uid,
                "email": user_record.email,
                "first_name": custom_claims.get("first_name", ""),
                "last_name": custom_claims.get("last_name", ""),
                "role": custom_claims.get("role", "user")
            }
        except Exception as e:
            logger.warning(f"Token verification failed: {e}")
            return None

    async def refresh_access_token(self, refresh_token: str) -> Optional[str]:
        """Issue new access token if refresh token is valid."""
        try:
            payload = jwt.decode(refresh_token, self.jwt_secret, algorithms=[self.jwt_algorithm])
            if payload.get("type") != "refresh":
                raise PermissionError("Invalid token type for refresh")

            user_id = payload.get("user_id")
            user_record = auth.get_user(user_id)

            return self._generate_access_token(user_id, user_record.email)
        except Exception as e:
            logger.warning(f"Token refresh failed: {e}")
            return None

    def _generate_access_token(self, user_id: str, email: str) -> str:
        payload = {
            "user_id": user_id,
            "email": email,
            "iat": datetime.utcnow(),
            "exp": datetime.utcnow() + self.access_token_expiry,
            "type": "access"
        }
        return jwt.encode(payload, self.jwt_secret, algorithm=self.jwt_algorithm)

    def _generate_refresh_token(self, user_id: str) -> str:
        payload = {
            "user_id": user_id,
            "iat": datetime.utcnow(),
            "exp": datetime.utcnow() + self.refresh_token_expiry,
            "type": "refresh"
        }
        return jwt.encode(payload, self.jwt_secret, algorithm=self.jwt_algorithm)


# Singleton instance
firebase_auth = FirebaseAuthService()

File: src/app/auth/firebaseAuth.js
import admin from 'firebase-admin';
import jwt from 'jsonwebtoken';
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

class FirebaseAuthService {
  constructor() {
    this._initializeFirebase();
    this.jwtSecret = process.env.JWT_SECRET || 'your-secret-key';
    this.jwtAlgorithm = 'HS256';
    this.accessTokenExpiry = 60 * 60; // 1 hour in seconds
    this.refreshTokenExpiry = 7 * 24 * 60 * 60; // 7 days in seconds
  }

  _initializeFirebase() {
    try {
      // Try to get Firebase credentials from environment
      const firebaseCredentials = process.env.FIREBASE_CREDENTIALS;
      if (firebaseCredentials) {
        const credDict = JSON.parse(firebaseCredentials);
        const cred = admin.credential.cert(credDict);
        admin.initializeApp({ credential: cred });
      } else {
        // Fallback to service account file
        const serviceAccountPath = process.env.FIREBASE_SERVICE_ACCOUNT_PATH;
        if (serviceAccountPath) {
          try {
            const serviceAccount = JSON.parse(
              readFileSync(serviceAccountPath, 'utf8')
            );
            const cred = admin.credential.cert(serviceAccount);
            admin.initializeApp({ credential: cred });
          } catch (error) {
            console.error('Error reading service account file:', error);
            // Try default credentials (for development)
            admin.initializeApp();
          }
        } else {
          // Use default credentials (for development)
          admin.initializeApp();
        }
      }
    } catch (error) {
      // If app already initialized, that's okay
      if (error.code !== 'app/already-initialized') {
        console.error('Firebase initialization error:', error);
        throw error;
      }
    }
  }

  async createUser(email, password, firstName, lastName) {
    try {
      const userRecord = await admin.auth().createUser({
        email: email,
        password: password,
        displayName: `${firstName} ${lastName}`,
        emailVerified: false,
      });

      // Set custom claims
      await admin.auth().setCustomUserClaims(userRecord.uid, {
        first_name: firstName,
        last_name: lastName,
        role: 'user',
      });

      return {
        id: userRecord.uid,
        email: userRecord.email,
        first_name: firstName,
        last_name: lastName,
        is_active: !userRecord.disabled,
        created_at: userRecord.metadata.creationTime,
      };
    } catch (error) {
      throw new Error(`Failed to create user: ${error.message}`);
    }
  }

  async signInUser(email, password) {
    try {
      // Note: Firebase Admin SDK doesn't have signInWithEmailAndPassword
      // In production, you'd use Firebase Auth REST API or Firebase Client SDK
      // For now, we'll get the user and generate tokens
      const userRecord = await admin.auth().getUserByEmail(email);

      if (userRecord.disabled) {
        throw new Error('User account is disabled');
      }

      // Generate JWT tokens
      const accessToken = this._generateAccessToken(userRecord.uid, userRecord.email);
      const refreshToken = this._generateRefreshToken(userRecord.uid);

      // Get custom claims
      const customClaims = userRecord.customClaims || {};

      return {
        access_token: accessToken,
        refresh_token: refreshToken,
        user: {
          id: userRecord.uid,
          email: userRecord.email,
          first_name: customClaims.first_name || '',
          last_name: customClaims.last_name || '',
          is_active: !userRecord.disabled,
          created_at: userRecord.metadata.creationTime,
        },
      };
    } catch (error) {
      throw new Error(`Authentication failed: ${error.message}`);
    }
  }

  async verifyToken(token) {
    try {
      // First try to verify as Firebase ID token
      try {
        const decodedToken = await admin.auth().verifyIdToken(token);
        const userRecord = await admin.auth().getUser(decodedToken.uid);
        const customClaims = userRecord.customClaims || {};

        return {
          uid: userRecord.uid,
          email: userRecord.email,
          first_name: customClaims.first_name || '',
          last_name: customClaims.last_name || '',
          role: customClaims.role || 'user',
        };
      } catch (firebaseError) {
        // If not a Firebase token, try as JWT
        const decoded = jwt.verify(token, this.jwtSecret, {
          algorithms: [this.jwtAlgorithm],
        });

        if (decoded.type === 'access') {
          // Get user from Firebase to get custom claims
          const userRecord = await admin.auth().getUser(decoded.user_id);
          const customClaims = userRecord.customClaims || {};

          return {
            uid: userRecord.uid,
            email: userRecord.email,
            first_name: customClaims.first_name || '',
            last_name: customClaims.last_name || '',
            role: customClaims.role || 'user',
          };
        }

        return null;
      }
    } catch (error) {
      console.error('Token verification failed:', error);
      return null;
    }
  }

  _generateAccessToken(userId, email) {
    const payload = {
      user_id: userId,
      email: email,
      exp: Math.floor(Date.now() / 1000) + this.accessTokenExpiry,
      iat: Math.floor(Date.now() / 1000),
      type: 'access',
    };
    return jwt.sign(payload, this.jwtSecret, { algorithm: this.jwtAlgorithm });
  }

  _generateRefreshToken(userId) {
    const payload = {
      user_id: userId,
      exp: Math.floor(Date.now() / 1000) + this.refreshTokenExpiry,
      iat: Math.floor(Date.now() / 1000),
      type: 'refresh',
    };
    return jwt.sign(payload, this.jwtSecret, { algorithm: this.jwtAlgorithm });
  }

  async refreshAccessToken(refreshToken) {
    try {
      const payload = jwt.verify(refreshToken, this.jwtSecret, {
        algorithms: [this.jwtAlgorithm],
      });

      if (payload.type !== 'refresh') {
        throw new Error('Invalid token type');
      }

      const userRecord = await admin.auth().getUser(payload.user_id);
      return this._generateAccessToken(payload.user_id, userRecord.email);
    } catch (error) {
      console.error('Token refresh failed:', error);
      return null;
    }
  }
}

// Global instance
export const firebaseAuth = new FirebaseAuthService();

